IPA-End-to-End.md

```diff
@@ -1924,6 +1927,80 @@ An example of an oblivious algorithm is the bitonic sorting algorithm [[1]](http
+
+In all cases, IPA provides the greatest benefit if there is a single system-wide (or user-wide) store of match keys.
+To facilitate this, support for storing and managing match keys at the level of the operating system provides the greatest utility.
+A system-wide utility will ensure that attribution can work across different apps, including browsing.
+Without a system-wide store, attribution can still work, though it will be limited to impressions and conversions in the same app.
+
+
+### Applications that can access web content
+
+If the embedded browser makes content accessible to the application, then the application is effectively acting as an independent browser.
+An application therefore needs distinct storage from the regular browser so that one application cannot access private information hosted by another application or browser.
+
+When IPA integrates into this sort of environment, the `getEncryptedMatchKey()` API needs to return an encryption of the match key that is specific to the combination of application and site.
+The same value cannot be used across different apps or sites, or the ciphertext that is provided might link activity across contexts.
+Integration of the IPA API at the system level therefore needs to segment any storage of _encrypted match keys_ based on the identity of the application that requests it, in addition to the site.
+
+Note that this is an imperfect model because activity on sites shown in a web view might be linkable to activity in browser or other apps.
+Many sites attempt to detect webviews and disable login, which can prevent more direct forms of linkability.
+However, some applications use embedded browsers to manage login and authorization of access to online services.
+Consequently, people can become accustomed to having to log in, despite that being a poor security practice, and in doing so create strong correlation between contexts.
+
+Aside from partitioning the storage of _encrypted match keys_ by app, no other adjustment is being considered.
+Sites can use match keys as usual.
+The app will be able to view or even change the encrypted match key, but the value will not reveal anything.
+This access might allow actions that could be seen to be malicious, such as:
+
+* An app can violate user privacy by _encrypted match keys_ to different sites that allow those sites to link user activity.
+
+* An app that captures an _encrypted match key_ from site content might be able to provide the value to helpers against the wishes of the site that is involved.
+
+* An app can provide sites with bogus _encrypted match keys_ or _encrypted match keys_ sourced from other contexts with a goal of spoiling attribution results.
+
+* An app might deny sites in the web view with access to the IPA APIs.
+
+* An app might misrepresent the identity of the site to the operating system, obtaining _encrypted match keys_ that are attributed to a site of its choice.
```

Comment:
I think you're right about this, but how would this work? Are we assuming that Webviews would be able to just call the system APIs directly? Or are we assuming that it would inject JavaScript into the page in some fashion to achieve this? Just walk me through how one would implement this attack step by step.

Reply:
This is a bit of a stretch, but you can imagine that the OS provides an API that takes the application identity (which might be implicit) and the identity of the site.  The web view is generally responsible for invoking the API when it loads a site.  The OS more or less has to trust that the code invoking the API is telling the truth about the site identity.  For a trustworthy browser, this isn't a big deal, but a malicious app that embeds a webview is not trustworthy.

In the general case, apps will embed the webview and the webview code will do the right thing.  After all, that code is generally provided by a trustworthy browser.  But the webview is just code that is run by the app, so the "webview" might be a wholly new browser, implemented from the ground up by the malicious app.  That's a bit extreme and probably bloats the app binary more than they might like (small payload = more likely to be accepted), so the malicious app might be able to modify the operation of the webview to access the API.

There are probably ways in which this can be secured, such as by insisting that webviews that access the API operate as distinct processes with different privileges, but these techniques are non-trivial and would not be user-distinguishable.  And webview code is complex, so it is possible that a malicious app could trick it into using the wrong site identity.

To be clear, I think that the amount of effort we put in to secure people from this sort of attack doesn't need to extend to this sort of defense, but it is worth describing this sort of attack.